### DISCUSSION TITLE
Units for viscosity?

### DISCUSSION THREAD

**Discussion by arjun-mani** [InteractiveComputerGraphics/SPlisHSPlasH#170] [Q&A]
Hi, thanks for this great simulator. I had a couple of quick questions on the viscosity method, assuming we were to use the 'standard' method:

1. Does the model take in the dynamic or kinematic viscosity? The docs seem to say kinematic viscosity, but I took a quick look at the source code and it seemed to treat the input as dynamic viscosity. I could be wrong here, so wanted to confirm.

2. What are the units for the viscosity? In other words - if I wanted to simulate water, what value with what units should I input?

Thanks very much in advance for your help!

**Comment by janbender** [comment]
Hi, 

1. It is the kinematic viscosity. Maybe the code is a bit confusing due to optimizations but I just checked it again. 
2. The units are the standard units (see also here: https://en.wikipedia.org/wiki/Viscosity): m^2/s

Hope that helps!

**Reply by arjun-mani** [reply]
Hi @janbender, thanks very much for your reply. I guess what confuses me is that I have set v=0.000001 (which is close to kinematic viscosity of water), and also v=0.00008 (which is close to honey), and I see very little difference between the simulations (using DamBreak.json). Moreover, v=0.01 still seems to produce low-viscous behavior (despite being extremely high for kinematic viscosity). I'm wondering whether you could clarify or recommend values that I could use to generate different viscous behavior. Thanks again for your help!

**Comment by janbender** [comment]
Hi, sorry for the late answer. We just double checked that it is the kinematic viscosity. However, note that the coefficient is dependent on the scale of your simulation since it has the unit m^2/s. For example, if I simulate the scene CoilingModel_Weiler2018.json which should show the behavior of a honey like material, I used a viscosity coefficient of 5 m^2/s. This is quite high. However, if you take a closer look in the scene file, the honey model at the beginning is 9 meter high. If I scale down the scene to a realistic size, a realistic coefficient for honey can be used. I took 0.01 m^2/s (see https://en.wikipedia.org/wiki/List_of_viscosities). Here is the adapted scene file: 

[honey.zip](https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/files/8814852/honey.zip)

**Reply by NiteeshMidlagajni** [reply]
Hello,
Could you please specify the values for the following parameters that would best approximate stable water properties?
-particleRadius
-viscosity
-timeStepSize (included here because I observe different behavior when step size is different while the other params remain constant)

If I use realistic values of 0.0025 for size, 0.000001 for viscosity and 0.001 as stepsize, the particles which are contained in an animated body jump around and 'spill' out of it while the body is moving. I have to heavily increase the viscosity(0.01) and decrease step size(0.0002) in order to for particles to not jump and remain contained within the moving animated body.

**Comment by janbender** [comment]
The particleRadius depends on the dimension of you scene. E.g., if you want to simulate 100m of a river, then you should not tkae a size of 0.0025 since you will get far too many particles. The particle size defines the resolution of your particle sampling. 

Concerning viscosity: either you take a slightly higher viscosity or you have to use a filter like XSPH, otherwise you can get chaotic particle motion. Many SPH simulators use XSPH which is an established method. 

If you have an animated body, you should be careful, since it can move quickly and this would require a smaller step size. I would recommend to take adaptive time steps using the CFL condition.
